src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Localization;
+using OrchardCore.ContentLocalization.Services;
+using OrchardCore.DisplayManagement.ModelBinding;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Localization;
+using OrchardCore.Modules;
+
+namespace OrchardCore.ContentLocalization.Controllers
+{
+    [Feature("OrchardCore.ContentLocalization.ContentCulturePicker")]
+    public class ContentCulturePickerController : Controller, IUpdateModel
+    {
+        private readonly ILocalizationService _locationService;
+        private readonly IContentCulturePickerService _culturePickerService;
+        private readonly string _tenantPrefix;
+
+        public IHtmlLocalizer T { get; }
+
+        public ContentCulturePickerController(
+            IHtmlLocalizer<AdminController> localizer,
+            ILocalizationService locationService,
+            IContentCulturePickerService culturePickerService,
+            ShellSettings shellSettings
+           )
+        {
+            T = localizer;
+            _locationService = locationService;
+            _tenantPrefix = "/" + shellSettings.RequestUrlPrefix;
+            _culturePickerService = culturePickerService;
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> RedirectToLocalizedContent(string targetCulture, PathString contentItemUrl)
+        {
+            // Invariant culture name is empty so a null value is bound.
+            targetCulture = targetCulture ?? "";
+
+            if (contentItemUrl == "/Error")
+            {
+                return NoContent();
+            }
+
+            if (!contentItemUrl.HasValue)
+            {
+                contentItemUrl = "/";
+            }
+
+            var supportedCultures = await _locationService.GetSupportedCulturesAsync();
+
+            if (!supportedCultures.Any(t => t == targetCulture))
+            {
+                return NoContent();
+            }
+
+            // Redirect the user to the Content with the same localizationSet as the ContentItem of the current url
+            var localizations = await _culturePickerService.GetLocalizationsFromRouteAsync(contentItemUrl);
+            if (localizations.Any())
+            {
+                var localization = localizations.SingleOrDefault(l => String.Equals(l.Culture, targetCulture, StringComparison.OrdinalIgnoreCase));
+
+                if(localization is object)
```

Comment:
If I ever see this line again, I remove the permissions ;)

Reply:
`if (localization != null)`

Reply:
When i saw this, i thought about Nick ;)

Reply:
Whats wrong with this!!! I even got it added to SonarQube!

Reply:
It's a micro optimization that doesn't add any value in terms of perf and makes the code more complex to read.